### ISSUE TITLE
Priority Sampling Formula

### ISSUE THREAD

**Issue by nemtiax** [werner-duvaud/muzero-general#42]
In Appendix G of the muzero paper, they define the priority of a sample as p_i = | nu_i - z_i |, and write "nu is the search value and z the observed n-step return."  (I'll use "nu" in place of ν for clarity when comparing with v)

However, in self_play.py, line 335, you seem to calculate priority as | v_i - nu_i |.

There are three related but distinct quantities here:

v - the output of the value head for a position
nu (ν) - the value estimate returned by MCTS for a position
z - the bootstrapped value of a position calculated using the next k observed rewards and the nu value k steps in the future, and discounting appropriately.

See Section 3 of the paper for definitions of these three terms.

It seems to me that the code currently does not match the priority calculation in the paper.  Is this intentional?  The | v - nu | formulation in the code has the advantage that priorities can be quickly updated whenever a position is used in training, because a new v is obtained.  The | nu - z | version in the paper is not amenable to priority updates, because that would seem to require at a minimum re-running MCTS to re-estimate nu, and at most re-playing the game for k round to re-estimate z, both of which are products of the current weights.

**Comment by werner-duvaud** [comment]
Hi,

Yes there are questions about that and we are not sure to doing it right.
The paper uses ν (the search value) and z (the observed n-step return).
But to update the priorities during the training, I think we would have to do the MCTS search at every training step for every step in the batch. To avoid that, we use the value predicted by the network instead of the search value. See https://github.com/werner-duvaud/muzero-general/blob/dfec4079f5bad7b3b3ffcb47e98a1298e2c3a03a/trainer.py#L216
For the initial priority, you can use the max recorded priority with the setting in the config or we tried to propose a priority with the available information at the current time step: the value predicted by the network and the search value.
We thought again about it and finally, we have changed the initial priority by waiting the end of the episode, now it matches the paper formula, but we will keep the different formula (see above) for the updated priorities.
